protocol_rfcs/type-widening.md

```diff
@@ -0,0 +1,143 @@
+Field Name | optional/required | Description
+-|-|-
+`tableVersion`| required | The version of the table when the type change was applied.
+`fromType`| required | The type of the column or field before the type change.
+`toType`| required | The type of the column or field after the type change.
+`fieldPath`| optional | When updating the type of a map key/value or array element only: the path from the struct field holding the metadata to the map key/value or array element that was updated.
+
+The `fieldPath` value is "key", "value" and "element"  when updating resp. the type of a map key, map value and array element.
+The `fieldPath` value for nested maps and nested arrays are prefixed by their parents's path, separated by dots.
+
+The following is an example for the definition of a column that went through two type changes:
+```json
+{
+    "name" : "e",
+    "type" : "long",
+    "nullable" : true,
+    "metadata" : { 
+      "delta.typeChanges": [
+        {
+          "tableVersion": 1,
+          "fromType": "short",
+          "toType": "int"
+        },
+        {
+          "tableVersion": 5,
+          "fromType": "int",
+          "toType": "long"
+        }
+      ]
+    }
+  }
+```
+
+The following is an example for the definition of a column after changing the type of a map key:
+```json
+{
+    "name" : "e",
+    "type" : {
+      "type": "map",
+      "keyType": "double",
+      "valueType": "int",
+      "valueContainsNull": true
+    },
+    "nullable" : true,
+    "metadata" : { 
+      "delta.typeChanges": [
+        {
+          "tableVersion": 2,
+          "fromType": "float",
+          "toType": "double",
+          "fieldPath": "key"
+        }
+      ]
+    }
+  }
+```
+
+The following is an example for the definition of a column after changing the type of a map value nested in an array:
+```json
+{
+    "name" : "e",
+    "type" : {
+      "type": "array",
+      "elementType": {
+        "type": "map",
+        "keyType": "string",
+        "valueType": "decimal(10, 4)",
+        "valueContainsNull": true
+      },
+      "containsNull": true
+    },
+    "nullable" : true,
+    "metadata" : { 
+      "delta.typeChanges": [
+        {
+          "tableVersion": 2,
+          "fromType": "decimal(6, 2)",
+          "toType": "decimal(10, 4)",
+          "fieldPath": "element.key"
+        }
+      ]
+    }
+  }
+```
+
+## Writer Requirements for Type Widening
+
+When Type Widening is enabled (when the table property `delta.enableTypeWidening` is set to `true`), then:
+- Writers should allow updating the table schema to apply an **allowed type change** to a column, struct field, map key/value or array element.
+- Writers must record type change information in the `metadata` of the nearest ancestor [StructField](#struct-field). See [Type Change Metadata](#type-change-metadata).
+
+When Type Widening is supported (when the `writerFeatures` field of a table's `protocol` action contains `enableTypeWidening`), then:
```

Comment:
Also, the writes must _preserve_ the `delta.typeChanges` field in the `metadata` fields in the schema when a schema is updated.

Do we also want to include limits to the preservation requirement? E.g. the `delta.typeChanges` could be garbage collected when the changes refer to commits that are no longer present in the table's history.

Reply:
I included a requirement on preserving the delta.typeChanges field and when writers are allowed to drop it.

And yes, dropping the field is possible as soon as the current version doesn't have any file that were written before the type change (I was initially too strict in the design doc)